### PR TITLE
QA2 - ES relevance improvements and query generation refactor

### PR DESCRIPTION
**Pull request by nonword**
This is a large PR. It includes two major efforts:

1. Updated ES queries for the new v8 index
2. Refactor how search queries are built

If you want to see the result of the first effort, [this diff](https://github.com/NYPL/discovery-api/compare/main..qa2-prerefactor?expand=1) shows just the parts that are mainly about updating the client to work with the v8 index, adding an assessment script, and doing a rough implementation of the updated queries.

This PR builds on top of that first effort to also attempt to clarify search query generation. It adds simple tools for query construction so that the important decisions about matching are foregrounded, with the fiddly ES query generation happening in the background. The aim is to make it much clearer what matches are active for a given query and scope and make it clear how to change things.

Also: This includes a new `buildingLocation` agg and support for `filters[buildingLocation]=...`

https://newyorkpubliclibrary.atlassian.net/browse/SCC-4312

### CODE REVIEW

File: swagger.v1.1.x.json
Comment (charmingduchess) [inline swagger.v1.1.x.json:1]: thanks for backfilling this documentation.

File: test/elastic-query-builder.test.js
Comment (charmingduchess) [inline test/elastic-query-builder.test.js:1]: These are great and actually go a long way in documenting how we're using ES. 🤘🏻

File: test/elasticsearch-utils.test.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      expect(esUtils.escapeQuery('Arkheologii︠a︡ Omska : illi︠u︡strirovannai︠a︡ ėnt︠s︡iklopedii︠a︡')).to.equal('Arkheologii︠a︡ Omska \\: illi︠u︡strirovannai︠a︡ ėnt︠s︡iklopedii︠a︡')
+    })
+
+    it('should escape colons in hyphenated phrases', function () {
+      expect(esUtils.escapeQuery('Arkheologii︠a︡ Omska : illi︠u︡strirovannai︠a︡ ėnt︠s︡iklopedii︠a︡ / Avtor-sostavitelʹ: B.A. Konikov.')).to.equal('Arkheologii︠a︡ Omska \\: illi︠u︡strirovannai︠a︡ ėnt︠s︡iklopedii︠a︡ \\/ Avtor\\-sostavitelʹ\\: B.A. Konikov.')
+    })
+  })
+
+  describe('prefixMatch', () => {
+    it('builds a prefix clause', () => {
```
Comment (charmingduchess) [inline test/elasticsearch-utils.test.js:32]: i think these could use a test where there is no boost

File: test/resources.test.js [hunk 2]
```diff
@@ -219,21 +145,7 @@ describe('Resources query', function () {
...
-                    terms: {
-                      nyplSource: ['recap-hl']
-                    }
-                  }
-                }
-              }
-            ]
-          }
-        })
+        expect(body).to.nested.include({ 'query.bool.filter[0].bool.must_not.terms.nyplSource[0]': 'recap-hl' })
```
Comment (charmingduchess) [inline test/resources.test.js:148]: would this kind of assertion have worked with our previous code? or did the code change so now we can do this check?

File: test/resources.test.js [hunk 3]
```diff
@@ -573,27 +477,27 @@ describe('Resources query', function () {
...
     it('should ignore check in card items when merge_checkin_card_items is falsey', () => {
       expect(resourcesPrivMethods.itemsQueryContext({ merge_checkin_card_items: false }))
-        .to.deep.equal({ must_not: [{ exists: { field: 'items.electronicLocator' } }, { term: { 'items.type': 'nypl:CheckinCardItem' } }] })
+        .to.deep.equal({ must_not: [{ term: { 'items.type': 'nypl:CheckinCardItem' } }] })
     })
 
     it('should include check in card items when merge_checkin_card_items is truthy', () => {
       expect(resourcesPrivMethods.itemsQueryContext({ merge_checkin_card_items: true }))
-        .to.deep.equal({ must_not: [{ exists: { field: 'items.electronicLocator' } }] })
+        .to.deep.equal({ must: { match_all: {} } })
```
Comment (charmingduchess) [inline test/resources.test.js:490]: This and the next two tests don't obviously assert what the `it` string says. To be fair, it doesn't look like they did before either. Maybe some comments would help explain? In particular the test on line 498 calls out electronic items, but doesn't appear to assert their absence in anyway.

**Review by charmingduchess — COMMENTED** [review]
Tremendous work! Most of my comments are just for clarification and some comments/documentation requests. I'm excited about the elastic search request encapsulation in particular.

File: lib/api-request.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  **/
+  hasScope (scopes) {
+    return scopes.includes(this.params.search_scope)
+  }
+
+  /**
+  * Returns true if query is fully wrapped in quotes
+  **/
+  queryIsFullyQuoted () {
+    return /^"[^"]+"$/.test(this.params.q)
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/api-request.js:49]: Good call. Added to the next PR

File: lib/elasticsearch/config.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    fields: [
+      'title^5',
+      'title.folded^2',
+      'titleAlt.folded',
+      'uniformTitle.folded',
+      'titleDisplay.folded',
+      'seriesStatement.folded',
+      'contentsTitle.folded',
+      'donor.folded',
+      'parallelTitle.folded^5',
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/elasticsearch/config.js:46]: We specially boost non-folded title just to give latin-based text matches with diacritics an edge over similar matches that match with diacritics removed. The "folding_analyzer" analyzer we're using uses the icu tokenizer and folder. I'm not sure what folding the folder does on non-latin languages, but the tokenizer is key as it knows how to break up cjk, etc into appropriate terms. So until we find otherwise, I think it's sufficient to match parallel values on just the .folded field

File: lib/elasticsearch/config.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    fields: ['subjectLiteral^2', 'subjectLiteral.folded', 'parallelSubjectLiteral.folded']
+  },
+  series: {
+    fields: ['seriesStatement.folded']
+  },
+  callnumber: {
+    fields: null
+  },
+  standard_number: {
+    fields: null
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/elasticsearch/config.js]: Historically, all search-scopes were configured in this const - with an array of `fields` to use in the `query_string`/multi-match query. With ES search improvements, two of these search-scopes don't use multi-match at all, so there's no need for a central config. However, this const remains a central, authoritative config for what search-scopes are available. I think future refactoring may change this, but for now I've just removed the `fields` prop from these search-scopes and added a comment to clarify

File: test/elasticsearch-utils.test.js (hunk 1 above)
Comment (nonword) [reply test/elasticsearch-utils.test.js:32]: Added

File: lib/elasticsearch/elastic-query-builder.js [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+        })
+      })
+
+    this.query.addMust(should.length === 1 ? should[0] : { bool: { should } })
+  }
+
+  /**
+  * Examine request for user-filters. When found, add them to query.
+  */
+  applyFilters () {
```
Comment (charmingduchess) [inline lib/elasticsearch/elastic-query-builder.js:450]: Can this method be broken down any more?

File: test/resources.test.js (hunk 2 above)
Comment (nonword) [reply test/resources.test.js:148]: I think these "nested" assertions have been available for a while. They do help where you want to make a value assertion about a specific nested property and are okay with other aspects of the object changing

File: test/resources.test.js (hunk 3 above)
Comment (nonword) [reply test/resources.test.js:490]: Good catch. Thanks for reading closely. I looked more closely at these tests. A couple were using config we no longer support. I updated the `it` statement to better reflect what we're testing.

File: lib/availability_resolver.js
[the remark this answers is not in the document]
Comment (nonword) [reply lib/availability_resolver.js:1]: You mean the new `bnum` check? This was an old staged commit that got brought into this work. I believe I encountered it in a no-results case. Essentially, if there are no results, `_fixItemStatusAggregation` should return early. Previously, it failed on the next call (because `bnum` was `undefined`). I suspect this is a theoretical edge case (or only comes up in practice in unit tests), but the function definitely needs that `bnum` to be set to proceed, so I'd like to keep the check.

File: lib/elasticsearch/elastic-query-builder.js (hunk 7 above)
Comment (nonword) [reply lib/elasticsearch/elastic-query-builder.js:450]: I agree it should be, but this specific code is unchanged from previous form (just moved). So I'd like to keep that as a project for a future refactor if that's alright.

File: lib/jsonld_serializers.js [hunk 8]
```diff
@@ -256,6 +261,9 @@ class ResourceSerializer extends JsonLdItemSerializer {
       }
     }
 
+    // DFE depends on this being set to an empty array when null:
+    stmts.electronicResources = stmts.electronicResources || []
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/jsonld_serializers.js:265]: Yes, but as it's almost entirely deprecated I think it's simplest to keep this shim in place until DFE is entirely removed. It's no longer serving Bib Details pages, but it's possible we may still need to refer to it up until it's entirely gone.

File: lib/response_massager.js [hunk 9]
```diff
@@ -30,14 +30,14 @@ class ResponseMassager {
         hit._source.items = hit.inner_hits.items.hits.hits.map((itemHit) => itemHit._source)
 
         // Grab the "items" inner_hits total as numItemsMatched:
-        hit._source.numItemsMatched = [hit.inner_hits.items.hits.total]
+        hit._source.numItemsMatched = [hit.inner_hits.items.hits.total.value]
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/response_massager.js:33]: ES v8 doesn't return exact counts after 10K by default, so `..hits.total` is an object like:
```
    "total": {
      "value": 10000,
      "relation": "gte"
    },
```
(The `relation` is "eq" for counts < 10K)

File: scripts/assess-query-targets.js [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  return Object.keys(goodHits)
+    .reduce((h, expectedIndex) => {
+      const bibid = goodHits[expectedIndex].bibid
+      if (!bibid) {
+        console.warn('No "good hits" defined?')
+        return h
+      }
+      const actualIndex = results.findIndex((result) => result.result.uri === bibid)
+      // Consider it a pass if within 5 of expected index (adjusting for number of good hits)
+      const pass = actualIndex >= 0 && Math.abs(actualIndex - expectedIndex) <= (5 + goodHits.length)
```
[the remark this answers is not in the document]
Comment (nonword) [reply scripts/assess-query-targets.js]: I refactored this a bit

File: lib/elasticsearch/utils.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        query: clause
+      }
+    }
+  }
+
+  return clause
+}
+
+/**
+* Given a field name and a value, returns a plainobject representing a ES
```
[the remark this answers is not in the document]
Comment (nonword) [reply lib/elasticsearch/utils.js:78]: Added

File: scripts/assess-query-targets.js
[the remark this answers is not in the document]
Comment (nonword) [reply scripts/assess-query-targets.js:1]: I made the documentation get a little more quickly to the important specifics.

File: scripts/assess-query-targets.js [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+      if (report.document) {
+        reportError(`  "${report.document.title[0]}" at index ${report.actualIndex}`)
+      } else {
+        reportError(`  ${report.bibid} not found in results`)
+      }
+    })
+  }
+
+  // Summarize results:
+  reportOn(`${pass.length} out of ${query.orderedHits.length} PASS`, ratioPass === 1 ? 'success' : (ratioPass < 0.5 ? 'failure' : 'mixed'))
```
[the remark this answers is not in the document]
Comment (nonword) [reply scripts/assess-query-targets.js]: Refactored a bit

**Review by charmingduchess — APPROVED** [review]
